src/MatrixFields/matrix_field_utils.jl

```diff
@@ -0,0 +1,141 @@
+function banded_matrix_info(field)
+    space = axes(field)
+    field_ld, field_ud = outer_diagonals(eltype(field))
+
+    # Find the diagonal index of the value that ends up in the bottom-right
+    # corner of the matrix, as well as the amount by which the field's diagonal
+    # indices get shifted when it is converted into a matrix.
+    bottom_corner_matrix_d, matrix_d_minus_field_d = if field_ld isa PlusHalf
+        if space.staggering isa Spaces.CellCenter
+            1, half # field is a face-to-center matrix
+        else
+            -1, -half # field is a center-to-face matrix
+        end
+    else
+        0, 0 # field is either a center-to-center or face-to-face matrix
+    end
+
+    n_rows = Spaces.nlevels(space)
+    n_cols = n_rows + bottom_corner_matrix_d
+    matrix_ld = field_ld + matrix_d_minus_field_d
+    matrix_ud = field_ud + matrix_d_minus_field_d
+    matrix_ld <= 0 && matrix_ud >= 0 ||
+        error("BandedMatrices.jl does not yet support matrices that have \
+               diagonals with indices in the range $matrix_ld:$matrix_ud")
+
+    return n_rows, n_cols, matrix_ld, matrix_ud
+end
```

Comment:
```suggestion
# Find the diagonal index of the value that ends up in the bottom-right
# corner of the matrix, as well as the amount by which the field's diagonal
# indices get shifted when it is converted into a matrix.
bottom_corner_matrix_d(::Spaces.Staggering, ::Int) = 0          # field is either a center-to-center or face-to-face matrix
bottom_corner_matrix_d(::Spaces.CellCenter, i::PlusHalf) = 1    # field is a face-to-center matrix
bottom_corner_matrix_d(::Spaces.CellFace, i::PlusHalf) = -1     # field is a center-to-face matrix
matrix_d_minus_field_d(::Spaces.Staggering, ::Int) = 0          # field is either a center-to-center or face-to-face matrix
matrix_d_minus_field_d(::Spaces.CellCenter, i::PlusHalf) = half # field is a face-to-center matrix
matrix_d_minus_field_d(::Spaces.CellFace, i::PlusHalf) = -half  # field is a center-to-face matrix

function banded_matrix_info(field)
    field_ld, field_ud = outer_diagonals(eltype(field))
    space = axes(field)
    staggering = space.staggering
    n_rows = Spaces.nlevels(space)
    n_cols = n_rows + bottom_corner_matrix_d(staggering, field_ld)
    matrix_ld = field_ld + matrix_d_minus_field_d(staggering, field_ld)
    matrix_ud = field_ud + matrix_d_minus_field_d(staggering, field_ld)
    matrix_ld <= 0 && matrix_ud >= 0 ||
        error("BandedMatrices.jl does not yet support matrices that have \
               diagonals with indices in the range $matrix_ld:$matrix_ud")
    return n_rows, n_cols, matrix_ld, matrix_ud
end
```
I might break that double union splitting into dispatch, I think it maybe reads a bit easier. (please double check my work there)

Reply:
Ok, yeah, I agree the current version is hard to read. Will do!